src/config.rs

```diff
@@ -0,0 +1,136 @@
+use ipc_sdk::subnet_id::SubnetID;
+use serde::de::{Error, SeqAccess};
+use serde::{Deserialize, Deserializer};
+use url::Url;
+
+/// The top-level struct representing the config. Calls to [`Config::from_file`] deserialize into this
+/// struct.
+#[derive(Deserialize)]
+pub(crate) struct Config {
+    pub subnets: HashMap<String, Subnet>,
+}
+
+/// Represents a subnet declaration in the config.
+#[derive(Deserialize)]
+pub struct Subnet {
+    #[serde(deserialize_with = "deserialize_path")]
+    path: SubnetID,
+    rpc_api: Url,
+    auth_token: Option<String>,
+    #[serde(deserialize_with = "deserialize_accounts", default)]
+    accounts: Vec<Address>,
+}
+
+impl Config {
+    /// Reads a TOML configuration file specified in the `path` and returns a [`Config`] struct.
+    pub fn from_file(path: &str) -> Result<Self> {
+        let contents = fs::read_to_string(path)?;
+        let config: Config = toml::from_str(&contents)?;
+        Ok(config)
+    }
+}
+
+/// A serde deserialization method to deserialize a subnet path string into a [`SubnetID`].
+fn deserialize_path<'de, D>(deserializer: D) -> Result<SubnetID, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    struct SubnetIDVisitor;
+    impl<'de> serde::de::Visitor<'de> for SubnetIDVisitor {
+        type Value = SubnetID;
+
+        fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
+            formatter.write_str("a string")
+        }
+
+        fn visit_str<E>(self, v: &str) -> std::result::Result<Self::Value, E>
+        where
+            E: Error,
+        {
+            SubnetID::from_str(v).map_err(E::custom)
+        }
+    }
+    deserializer.deserialize_str(SubnetIDVisitor)
+}
+
+/// A serde deserialization method to deserialize a list of account strings into a vector of
+/// [`Address`].
+fn deserialize_accounts<'de, D>(deserializer: D) -> Result<Vec<Address>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    struct AddressSeqVisitor;
+    impl<'de> serde::de::Visitor<'de> for AddressSeqVisitor {
+        type Value = Vec<Address>;
+
+        fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
+            formatter.write_str("a sequence of strings")
+        }
+
+        fn visit_seq<A>(self, mut seq: A) -> std::result::Result<Self::Value, A::Error>
+        where
+            A: SeqAccess<'de>,
+        {
+            let mut vec: Vec<Address> = Vec::new();
+            while let Some(value) = seq.next_element::<String>()? {
+                let a = Address::from_str(value.as_str()).map_err(Error::custom)?;
+                vec.push(a);
+            }
+            Ok(vec)
+        }
+    }
+    deserializer.deserialize_str(AddressSeqVisitor)
+}
+
+#[cfg(test)]
+mod tests {
+    use std::str::FromStr;
+
+    use fvm_shared::address::Address;
+    use ipc_sdk::subnet_id::SubnetID;
+    use url::Url;
+
+    use crate::config::Config;
+
+    #[test]
+    fn read_example_config() {
+        let path = "config/test_example.toml";
```

Comment:
This path `config/test_example.toml` is relative, if the tests are triggered from other paths but the repo root, then tests will break. Using the file's absolute path as base would be safer for tests. But still, `test_example.toml` could potentially be removed, then tests will break as well. Or we can dynamically create temp files and destroy them, which will be neater.
I think we can just implement `From<String> for Config` and then Config::from_file() calls Config::from(string), and in tests, there is no need to depend on external files.

Reply:
Yes, thank you for bringing that up. I like this idea and it would make the test much more hermetic. The only downside is we won't be calling `Config::from_file` in a test, but I suppose we can live with that.